src/App.tsx

```diff
@@ -1,26 +1,279 @@
+  const [editTitle, setEditTitle] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const editInputRef = React.useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage('Unable to load todos'));
+  }, []);
+
+  useEffect(() => {
+    if (!tempTodo) {
+      inputRef.current?.focus();
+    }
+  }, [tempTodo]);
+
+  useEffect(() => {
+    if (editId) {
+      editInputRef.current?.focus();
+    }
+  }, [editId]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  let visibleTodos = todos;
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const normalizedTitle = title.trim();
+
+    if (!normalizedTitle) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setTempTodo({
+      id: 0,
+      title: normalizedTitle,
+      completed: false,
+      userId: 0,
+    });
+
+    setErrorMessage('');
+
+    createTodo(normalizedTitle)
+      .then(todo => {
+        setTodos((prevTodos: Todo[]) => [...prevTodos, todo]);
+        setTitle('');
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => setTempTodo(null));
+  };
+
+  const handleSubmitEdit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const normalizedTitle = editTitle.trim();
+
+    if (!normalizedTitle) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setErrorMessage('');
+
+    updateTodo(editId, normalizedTitle)
+      .then(todo => {
+        setTodos((prevTodos: Todo[]) =>
+          prevTodos.map(prevTodo => {
+            if (prevTodo.id === todo.id) {
+              return {
+                ...prevTodo,
+                title: todo.title,
+              };
+            }
+
+            return prevTodo;
+          }),
+        );
+        setTitle('');
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setEditId(null);
+        setEditTitle('');
+      });
+  };
+
+  const handleCancelEdit = () => {
+    setEditId(null);
+    setEditTitle('');
+  };
+
+  const handleDeletingTodo = (id: number) => {
+    setErrorMessage('');
+    setProcessingIds(prevIds => [...prevIds, id]);
+
+    return deleteTodo(id)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+      })
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() =>
+        setProcessingIds(prevIds => prevIds.filter(prevId => prevId !== id)),
+      );
+  };
+
+  const handleClearCompleted = () => {
+    completedTodos.forEach(todo => handleDeletingTodo(todo.id));
+  };
+
+  useEffect(() => {
```

Comment:
move it up - to other useEffect hook calls - it's easier to read it, if component has structure